gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+	int buttonHeight = h / 3;
+
+	PicButtonWidget *playButton = new PicButtonWidget(this, (buttonWidth / 3), buttonHeight / 3, (buttonWidth / 3) + buttonWidth * 2, buttonHeight, U32String("Play"), kPlayButtonCmd);
+	PicButtonWidget *loadButton = new PicButtonWidget(this, (buttonWidth / 3), (buttonHeight * 5) / 3, buttonWidth, buttonHeight, U32String("Saves"), kLoadButtonCmd);
+	PicButtonWidget *editButton = new PicButtonWidget(this, buttonWidth + 2 * (buttonWidth / 3), (buttonHeight * 5) / 3, buttonWidth, buttonHeight, U32String("Edit"), kEditButtonCmd);
+	
+	playButton->useThemeTransparency(true);
+	loadButton->useThemeTransparency(true);
+	editButton->useThemeTransparency(true);
+
+	playButton->setGfxFromTheme("button_play.bmp", 0, false);
+	loadButton->setGfxFromTheme("button_load.bmp", 0, false);
+	editButton->setGfxFromTheme("button_options.bmp", 0, false);
+}
+
+void GridItemTray::handleCommand(CommandSender *sender, uint32 cmd, uint32 data) {
+	switch (cmd) {
+	case kPlayButtonCmd:
+		close();
+		sendCommand(kPlayButtonCmd, _entryID);
+		break;
+	case kLoadButtonCmd:
+		close();
+		sendCommand(kLoadButtonCmd, _entryID);
+		break;
+	case kEditButtonCmd:
+		close();
+		sendCommand(kEditButtonCmd, _entryID);
+		break;
+	case kCloseCmd:
+		close();
+		break;
+	default:
+		Dialog::handleCommand(sender, cmd, 0);
+		break;
+	}
+}
+
+void GridItemTray::handleMouseDown(int x, int y, int button, int clickCount) {
+	Dialog::handleMouseDown(x, y, button, clickCount);
+	if ((x < 0 || x > _w) || (y > _h || y < -(_grid->_gridItemHeight))) {
+		// Close on clicking outside
+		close();
+	} else if (y < 0 && clickCount >= 2) {
+		// Run on double clicking thumbnail
+		close();
+		sendCommand(kItemDoubleClickedCmd, _entryID);
+	}
+}
+
+void GridItemTray::handleMouseWheel(int x, int y, int direction) {
+	Dialog::handleMouseWheel(x, y, direction);
+	close();
+}
+
+void GridItemTray::handleMouseMoved(int x, int y, int button) {
+	Dialog::handleMouseMoved(x, y, button);
+	if ((x < 0 || x > _w) || (y > _h || y < -(_grid->_gridItemHeight))) {
+		// Close on going outside
+		close();
+	}
+}
+
+#pragma mark -
+
+Graphics::ManagedSurface *loadSurfaceFromFile(const Common::String &name) {
+	Graphics::ManagedSurface *surf = nullptr;
+	const Graphics::Surface *srcSurface = nullptr;
+	const String path = String::format("%s/%s", ConfMan.get("iconpath").c_str(), name.c_str());
+	if (name.hasSuffix(".png")) {
+#ifdef USE_PNG
+		Image::PNGDecoder decoder;
+		Common::FSNode fileNode(path);
+		Common::SeekableReadStream *stream = fileNode.createReadStream();
+		if (stream) {
+			if (!decoder.loadStream(*stream))
+				warning("Error decoding PNG");
+			
+			srcSurface = decoder.getSurface();
+			delete stream;
+			if (!srcSurface) {
+				warning("Failed to load surface : %s", name.c_str());
+			}
+			if (srcSurface && srcSurface->format.bytesPerPixel != 1) {
+				surf = new Graphics::ManagedSurface(srcSurface->convertTo(g_system->getOverlayFormat()));
+			}
+				
+		}
```

Comment:
What about else?

Reply:
I am actually unsure about this. I referenced the code of ThemeEngine::addBitmap(), and there was a check for format.bytesPerPixel != 1, but I have tried to generate PNGs with various settings, I have not been able to make a 1 bpp PNG, so I don't know if an else is even possible. I did remove the redundant `srcSurface &&` , in the second if statement to make it into an else if statement in 739cb8e.